api/api_ethereum.go

```diff
@@ -1086,77 +1190,126 @@ func (args *EthTransactionArgs) ToMessage(globalGasCap uint64, baseFee *big.Int,
-	return common.HexToHash("0x"), nil
+	if args.Nonce == nil {
+		// Hold the addresses mutex around signing to prevent concurrent assignment of
+		// the same nonce to multiple accounts.
+		api.publicTransactionPoolAPI.nonceLock.LockAddr(args.from())
+		defer api.publicTransactionPoolAPI.nonceLock.UnlockAddr(args.from())
+	}
+	if err := args.setDefaults(ctx, api.publicTransactionPoolAPI.b); err != nil {
+		return common.Hash{}, err
+	}
+	tx := args.toTransaction()
+	signedTx, err := api.publicTransactionPoolAPI.sign(args.from(), tx)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	// Check if signedTx is RLP-Encodable format.
+	_, err = rlp.EncodeToBytes(signedTx)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return submitTransaction(ctx, api.publicTransactionPoolAPI.b, signedTx)
 }
 
 // EthSignTransactionResult represents a RLP encoded signed transaction.
 // SignTransactionResult in go-ethereum has been renamed to EthSignTransactionResult.
 // SignTransactionResult is defined in go-ethereum's internal package, so SignTransactionResult is redefined here as EthSignTransactionResult.
 type EthSignTransactionResult struct {
 	Raw hexutil.Bytes `json:"raw"`
-	Tx  *Transaction  `json:"tx"`
-}
-
-// Transaction is an Ethereum transaction.
-type Transaction struct {
-	inner TxData    // Consensus contents of a transaction
-	time  time.Time // Time first seen locally (spam avoidance)
-
-	// caches
-	hash atomic.Value
-	size atomic.Value
-	from atomic.Value
-}
-
-// TxData is the underlying data of a transaction.
-//
-// This is implemented by DynamicFeeTx, LegacyTx and AccessListTx.
-type TxData interface {
-	txType() byte // returns the type ID
-	copy() TxData // creates a deep copy and initializes all fields
-
-	chainID() *big.Int
-	accessList() types.AccessList
-	data() []byte
-	gas() uint64
-	gasPrice() *big.Int
-	gasTipCap() *big.Int
-	gasFeeCap() *big.Int
-	value() *big.Int
-	nonce() uint64
-	to() *common.Address
-
-	rawSignatureValues() (v, r, s *big.Int)
-	setSignatureValues(chainID, v, r, s *big.Int)
+	Tx  *ethTxJSON    `json:"tx"`
 }
 
 // FillTransaction fills the defaults (nonce, gas, gasPrice or 1559 fields)
 // on a given unsigned transaction, and returns it to the caller for further
 // processing (signing + broadcast).
-func (api *EthereumAPI) FillTransaction(ctx context.Context, args EthTransactionArgs) (*EthSignTransactionResult, error) {
-	// TODO-Klaytn: Not implemented yet.
-	return nil, nil
+func (api *EthereumAPI) FillTransaction(ctx context.Context, args EthTransactionArgs) (*EthSignTransactionResult, error) { // Set some sanity defaults and terminate on failure
+	if err := args.setDefaults(ctx, api.publicTransactionPoolAPI.b); err != nil {
+		return nil, err
+	}
+	// Assemble the transaction and obtain rlp
+	tx := args.toTransaction()
+	data, err := tx.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	return &EthSignTransactionResult{Raw: data, Tx: formatTxToEthTxJSON(tx)}, nil
```

Comment:
This you check whether `data` has 0x78 EthereumTxTypeEnvolop prefix?

Reply:
Sorry for missing it.
I'll add handling logic for this.

Reply:
Acutally, i think in this case, `data` should not include `0x78` EthereumTxTypeEnvolop prefix.

Reply:
I think ethereumTypedTx encoding logic without `0x78` prefix has to be separate from TxHash function for this.!

Reply:
I agreed with @aidan-kwon @jimni1222.

Reply:
KEN Before:
```shell
curl http://localhost:8553 \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_fillTransaction","params":[{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x8c9f4468ae04fb3d79c80f6eacf0e4e1dd21deee", "value": "0x1", "nonce": "0x4", "gas": "0x9999", "maxPriorityFeePerGas": "0x5d21dba00"}],"id":1}'
```
Result - rawTx: `0x7802ef8302edaf048505d21dba008505d21dba00829999948c9f4468ae04fb3d79c80f6eacf0e4e1dd21deee0180c0808080`

KEN After:
```shell
curl http://localhost:8553 \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_fillTransaction","params":[{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x8c9f4468ae04fb3d79c80f6eacf0e4e1dd21deee", "value": "0x1", "nonce": "0x4", "gas": "0x9999", "gasPrice": "0x5d21dba00"}],"id":1}'
```
Result - rawTx: `0xe4048505d21dba00829999948c9f4468ae04fb3d79c80f6eacf0e4e1dd21deee0180808080`

Geth Result:
```shell
curl http://localhost:8545 \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_fillTransaction","params":[{"from": "0xca7a99380131e6c76cfa622396347107aeedca2d", "to": "0x8c9f4468ae04fb3d79c80f6eacf0e4e1dd21deee", "value": "0x1", "nonce": "0x4", "gas": "0x9999", "gasPrice": "0x5d21dba00"}],"id":1}'
```

Reply:
It seems that your "Before" test used TxTypeEthereumDynamic tx, but your "After" test used legacy type tx.
You may need some tricks to enable new tx types.

Reply:
@aidan-kwon 
I updated my comments. PTAL.

Reply:
I updated my test log correctly. PTAL.

Reply:
Thanks!